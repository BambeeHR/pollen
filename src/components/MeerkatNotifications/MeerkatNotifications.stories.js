import { select } from '@storybook/addon-knobs';
import BaseButton from '../BaseButton/BaseButton.vue';
import Notification from '../../constants/Notification';
import MeerkatNotifications from './MeerkatNotifications.vue';

const { Positions, Variants } = Notification;

export default {
  title: 'MeerkatNotifications',
  component: MeerkatNotifications,
};

export const Default = () => ({
  components: { BaseButton, MeerkatNotifications },
  props: {
    position: {
      default: select(
        'Position',
        Object.values(Positions),
        Positions.BOTTOM_CENTER
      ),
    },
  },
  data() {
    return {
      notifications: [
        {
          message: 'Lorem ipsum dolor sit amet',
          id: 1,
        },
        {
          message: 'Lorem ipsum dolor sit amet',
          variant: Variants.SECONDARY,
          id: 2,
          image: 'https://cdn.filestackcontent.com/enCa5BTVeIsLHcgkdpAc',
          showActions: false,
        },
        {
          message: 'Lorem ipsum dolor sit amet',
          variant: Variants.ERROR,
          id: 3,
        },
        {
          message: 'Lorem ipsum dolor sit amet',
          actions: [
            {
              label: 'Alert',
              handler: () => {
                alert('foo');
              },
            },
            {
              label: 'Close',
              handler: () => {
                this.remove(4);
              },
            },
          ],
          id: 4,
        },
      ],
    };
  },
  methods: {
    add() {
      const id = Math.max(...this.notifications.map((i) => i.id)) + 1;
      this.notifications.push({
        message: `I'm a new message. ID ${id}.`,
        id,
      });
    },
    remove(id) {
      this.notifications = this.notifications.filter((item) => item.id !== id);
    },
  },
  template: `
    <div>
      <BaseButton type="button" pre-icon="plus" @click="add">Add a notification</BaseButton>
      <MeerkatNotifications :notifications="notifications" :position="position" @close="remove" />
    </div>`,
});
