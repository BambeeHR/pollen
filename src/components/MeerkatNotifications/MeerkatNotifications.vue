<template>
  <transition-group
    v-if="notifications.length"
    tag="div"
    class="meerkat-notifications"
    :class="`meerkat-notifications--${position}`"
    name="meerkat"
    @before-leave="beforeLeave"
  >
    <MeerkatNotification
      v-for="{ message, actions = [], id, ...rest } in notifications"
      :key="id"
      v-bind="rest"
      @close="$emit('close', id)"
    >
      {{ message }}
      <template v-if="actions.length" #actions>
        <BaseButton
          v-for="{ label, handler } in actions"
          :key="label"
          :size="$options.Button.Sizes.SMALL"
          :variant="$options.Button.Variants.INVERTED"
          flat
          @click="handler"
        >
          {{ label }}
        </BaseButton>
      </template>
    </MeerkatNotification>
  </transition-group>
</template>

<script>
import BaseButton from '../BaseButton/BaseButton.vue';
import Button from '../../constants/Button';
import Notification from '../../constants/Notification';
import MeerkatNotification from '../MeerkatNotification/MeerkatNotification.vue';

/**
 * A container for displaying `MeerkatNotification` components. This can be
 * displayed on the top or bottom, in the center or right corner using the
 * `position` prop. It accepts a `notifications` array of objects. These objects
 * should have at the very least a `message` string and a unique `id` string or
 * number. Optionally, the notification object can have an array of `actions`
 * that have a string `label` and a `handler` function that is executed on
 * click. Any other attributes on the notification object will be transparently
 * passed through to the `MeerkatNotification` child.
 */
export default {
  Button,
  components: { BaseButton, MeerkatNotification },
  props: {
    /**
     * An array of notification objects. Should include at least a `message` and
     * `id`. An optional `actions` array of objects with `label` and `handler`
     * can also be passed in to override the default `x` close button of an
     * individual notification.
     */
    notifications: {
      type: Array,
      default: () => [],
    },
    /** One of `bottom-center`, `bottom-right`, `top-center`, or `top-right`. */
    position: {
      type: String,
      default: Notification.Positions.BOTTOM_CENTER,
      validator: (value) =>
        Object.values(Notification.Positions).includes(value),
    },
  },
  methods: {
    beforeLeave(el) {
      el.setAttribute('style', `top: ${el.offsetTop || 0}px`);
    },
  },
};
</script>

<style scoped>
.meerkat-notifications {
  @apply box-border fixed flex -my-2 p-4;
  max-width: 100vw;
  width: 25rem;
}

.meerkat-notifications .meerkat-notification {
  @apply duration-lazy my-2 transition-all;
}

.meerkat-notifications--top-center,
.meerkat-notifications--top-right {
  @apply flex-col-reverse top-0;
}

.meerkat-notifications--bottom-center,
.meerkat-notifications--bottom-right {
  @apply bottom-0 flex-col;
}

.meerkat-notifications--top-center,
.meerkat-notifications--bottom-center {
  @apply left-1/2 transform -translate-x-1/2;
}

.meerkat-notifications--top-right,
.meerkat-notifications--bottom-right {
  @apply right-0;
}

.meerkat-enter,
.meerkat-leave-to {
  opacity: 0;
}

.meerkat-notifications--bottom-center .meerkat-enter,
.meerkat-notifications--bottom-right .meerkat-enter,
.meerkat-notifications--bottom-center .meerkat-leave-to,
.meerkat-notifications--bottom-right .meerkat-leave-to {
  transform: translateY(100%);
}

.meerkat-notifications--top-center .meerkat-enter,
.meerkat-notifications--top-right .meerkat-enter,
.meerkat-notifications--top-center .meerkat-leave-to,
.meerkat-notifications--top-right .meerkat-leave-to {
  transform: translateY(-100%);
}

.meerkat-enter-active,
.meerkat-leave-active {
  z-index: -1;
}

.meerkat-enter-active {
  @apply ease-out;
}

.meerkat-leave-active {
  @apply absolute ease-in;
  width: calc(100% - 2rem);
}
</style>
