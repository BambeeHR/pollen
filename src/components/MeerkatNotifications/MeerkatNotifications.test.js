import { mount } from '@vue/test-utils';
import MeerkatNotifications from './MeerkatNotifications.vue';
import { Default } from './MeerkatNotifications.stories';

describe('MeerkatNotifications', () => {
  test('renders correctly', () => {
    const notifications = Default();
    const wrapper = mount(notifications);
    expect(wrapper.element).toMatchSnapshot();
  });
  test('does not render if there are no notifications passed', () => {
    const wrapper = mount(MeerkatNotifications);
    expect(wrapper.find('.meerkat-notifications').exists()).toBeFalsy();
  });
  // TODO(jon.jandoc): This is not rendering at all for some reason and I have
  // no idea why. Skipping for now.
  test.skip('renders an action', async () => {
    const action = jest.fn();
    const wrapper = mount(MeerkatNotifications, {
      propsData: {
        notifications: [
          {
            message: 'Lorem ipsum dolor sit amet',
            id: 1,
          },
          {
            message: 'Lorem ipsum dolor sit amet',
            actions: [
              {
                label: 'Close',
                handler: () => {
                  action();
                },
              },
            ],
            id: 2,
          },
        ],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.findAll('.meerkat-notification').length).toBe(2);
    await wrapper.findAll('button').trigger('click');
    expect(action).toHaveBeenCalled();
  });
});
