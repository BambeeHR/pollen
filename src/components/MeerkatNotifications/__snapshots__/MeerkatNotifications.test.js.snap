// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeerkatNotifications renders correctly 1`] = `
<div>
  <button
    class="base-button base-button--primary base-button--medium base-button--regular base-button--inline"
    type="button"
  >
    <div
      class="base-button__inner"
    >
      <svg
        aria-hidden="true"
        class="base-icon svg-inline--fa fa-plus fa-w-12 base-button__icon base-button__icon--pre"
        data-icon="plus"
        data-prefix="fal"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          d="M376 232H216V72c0-4.42-3.58-8-8-8h-32c-4.42 0-8 3.58-8 8v160H8c-4.42 0-8 3.58-8 8v32c0 4.42 3.58 8 8 8h160v160c0 4.42 3.58 8 8 8h32c4.42 0 8-3.58 8-8V280h160c4.42 0 8-3.58 8-8v-32c0-4.42-3.58-8-8-8z"
          fill="currentColor"
        />
      </svg>
      Add a notification
    </div>
  </button>
   
  <!---->
</div>
`;
